crypto/x509/x509_vfy.c

```diff
@@ -3112,6 +3085,11 @@ static int build_chain(X509_STORE_CTX *ctx)
             }
             x = sk_X509_value(ctx->chain, i-1);
 
+            /* Stop with success if current cert equals a trust anchor */
+            if (lookup_cert_match(ctx, x) != NULL
+                    && (trust = check_trust(ctx, num)) == X509_TRUST_TRUSTED)
+                break;
+
```

Comment:
I don't believe this new code is needed, or appropriate.  The exact match lookups are only appropriate for EE certs and already handled in check_trust() as fallback when partial chains are enabled.

Reply:
See my above comment to https://github.com/openssl/openssl/pull/13748#discussion_r555289273.

Reply:
Sorry, this code is not correct, because presumptively we have a chain that's not yet complete, and the very next line looks for an issuer of the most recently added untrusted certificate, that code also checks whether the newly added issuer is trusted or not.  Other than perhaps with an EE cert at depth 0, this check looks redundant, because the trust was checked when the cert was added to the chain and does not need to be checked again.

Reply:
Sorry, it **is** correct and even required because as mentioned there are situations where it does not even make sense to try looking for an issuer cert.

Meanwhile I've simplified the new code at this point to:
```
            if ((trust = check_trust(ctx, num)) == X509_TRUST_TRUSTED)
                break;
```
It is required because otherwise the following new test case `"accept ee-cert although deceptively matching also first cert in ee-certs"` would fail like this:
```
apps/openssl verify -auth_level 1 -purpose any -no_check_time -trusted test/certs/ee-noKID-certs.pem test/certs/ee-noKID-cert2.pem
C = US, ST = CA, L = Palo Alto, O = VMware, CN = nsxmanager.pks.vmware.local
error 18 at 0 depth lookup: self-signed certificate
error test/certs/ee-noKID-cert2.pem: verification failed
```
Moreover, the test case `"accept last-resort direct leaf match Ed25519-signed self-issued cert"`:
`verify -partial_chain -trusted test/certs/ee-ed25519.pem test/certs/ee-ed25519.pem`
would else produce the spurious intermittent error `X509_V_ERR_SIGNATURE_ALGORITHM_MISMATCH` just because the chain building needlessly tries to find an issuer cert, which is found in the trust store but cannot be used for signing.